accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -530,62 +577,282 @@ mixinModifier     ::= 'sealed' | 'base' | 'interface' | 'final'
-    with the previous rule, ensures you can't get a backdoor interface on a
-    mixin that doesn't want to expose one.*
+*   Implement the interface of a class or mixin marked `base`, `final` or
+    `sealed` outside of the library where it is declared.
+
+    ```dart
+    // a.dart
+    base class B {}
+    final class F {}
+    sealed class S {}
+
+    base mixin BM {}
+    final mixin FM {}
+    sealed mixin SM {}
+
+    // b.dart
+    import 'a.dart';
+
+    class C1 implements B {} // Error.
+    class C2 implements F {} // Error.
+    class C3 implements S {} // Error.
+
+    class C1 implements BM {} // Error.
+    class C2 implements FM {} // Error.
+    class C3 implements SM {} // Error.
+    ```
     
-*   Apply `mixin` to a class whose superclass is not `Object` or that declares a
-    _non-trivial generative constructor_.
-    *Such a class can have an `extends` clause of the form `extends Object`,
-    or no `extends` clause. It cannot have any `with` clause.*
+*   Mix in a mixin or mixin class marked `interface`, `final` or `sealed`
+    outside of the library where it is declared.
+
+    ```dart
+    // a.dart
+    interface mixin class I {}
+    final mixin class F {}
+    sealed mixin class S {}
+
+    interface mixin IM {}
+    final mixin FM {}
+    sealed mixin SM {}
+
+    // b.dart
+    import 'a.dart';
+
+    class C1 with I {} // Error.
+    class C2 with F {} // Error.
+    class C3 with S {} // Error.
+
+    class C1 with IM {} // Error.
+    class C2 with FM {} // Error.
+    class C3 with SM {} // Error.
+    ```
+
+A typedef can't be used to subvert these restrictions or any of the restrictions
+below. When extending, implementing, or mixing in a typedef, we look at the
+library where class or mixin the typedef resolves to is defined to determine if
+the behavior is allowed. *Note that the library where the _typedef_ is defined
+does not come into play. Typedefs cannot be marked with any of the new
+modifiers.*
+
+### Disallowing implementation
+
+It is a compile-time error if a subtype of a declaration is marked `base` or
+`final` is not marked `base`, `final`, or `sealed`. This restriction applies to
+both direct and indirect subtypes and along all paths that introduce subtypes:
+`implements` clauses, `extends` clauses, `with` clauses, and `on` clauses. This
+restriction applies even to types within the same library.
+
+*Once the ability to use as an interface is removed, it cannot be reintroduced
+in a subtype. If a class is marked `base` or `final`, you may still implement
+the class's interface inside the same library, but the implementing class must
+again be marked `base`, `final`, or `sealed` to avoid it exposing an
+implementable interface.*
+
+Further, while you can ignore some restrictions on declarations within the same
+library, you can't use that to ignore restrictions inherited from other
+libraries.
+
+We say a class or mixin declaration `D` *can't be implemented locally* if it
```

Comment:
Trying to grasp what this means, my current hypothesis is that: if `D` can't be implemented locally then it is a compile-time error for a `D2` to be declared in the same library as `D` if `D2` has a clause of the form `implements ... D ...` (and most likely also if it has `on ... D ...`).

Reply:
Correct for `implements ...D...`. No to the `on ... D ...`, as usual it's not a problem. 

The `on ... D ...` declared `mixin` will need to be `base`, `final` or `sealed` itself, and it too can't be implemented locally, so it cannot be implemented at all.

It can only come into play as the type of an *object* if it's applied to another type which implements `D`, in which case it inherits members from that class, which is what we want to ensure.

The crux is that we only do exhaustiveness on objects, not types.

```dart
sealed class A {}
// Assume these are only immediate subtypes of A
final class B extends A {}
final class C extends A {}
...
   List<A> l = ...;
   return switch (l) {
         List<B> lb => ...
         List<C> lb => ...
   }
```
This expression switch is not exhaustive, because `B` | `C` only exhaust *instances* of `A`.
It's possible to have a list of exactly type `List<A>`, even if `A` is abstract and cannot have any own instances.
We only care about exhaustiveness on instances, which is why `A` must be abstract and unimplementable, so all instance are instances of `B` or `C`.
We can allow other immediate subtypes, as long as those types are also abstract and uninhabitable by values that are not also `B` or `C`.
And mixins satisfy that.

An other-library declaration of
```dart
mixin M on A {
  ...
}
```
introduces a new direct subtype of `A`, but it doesn't introduce any possible instances that are not also subtypes of `B` or `C`. The mixin must be applied to a class which is *extensible* (`base` or no modifier) before an instance can be created.

Because of the other constraints, every extensible class which implements `A` is either declared in the same library as `A` and subtypes either `B` or `C` (if it directly subtypes `A`, it becomes part of the sealed set of subtypes, which we assume here to be just `B` and `C`).
`A` itself is not extensible from outside the library, and `M` is not implementable or instantiable.